File: Triangle/Javascript/solution_3.js
// https://leetcode.cn/problems/IlPe0q/
// https://www.lintcode.com/problem/109/

const minimumTotal = function (triangle) {
  let memo = new Map();
  return divide_conquer(triangle, 0, 0, memo);
};

function divide_conquer(triangle, x, y, memo) {
  if (x === triangle.length) {
    return 0;
  }

  let key = `${x}-${y}`;
  if (memo.has(key)) {
    return memo.get(key);
  }

  let left = divide_conquer(triangle, x + 1, y, memo);
  let right = divide_conquer(triangle, x + 1, y + 1, memo);
  let path = Math.min(left, right) + triangle[x][y];

  memo.set(key, path);
  return path;
}

let triangle = [[2], [3, 4], [6, 5, 7], [4, 1, 8, 3]];
console.log(minimumTotal(triangle));

File: Triangle/Rust/src/solution_3.rs
// https://leetcode.cn/problems/IlPe0q/
// https://www.lintcode.com/problem/109/

use std::{cmp::min, collections::HashMap};

pub fn minimum_total(triangle: Vec<Vec<i32>>) -> i32 {
    let mut map = HashMap::new();
    return divide_conquer(&triangle, 0, 0, &mut map);
}

fn divide_conquer(
    triangle: &Vec<Vec<i32>>,
    x: usize,
    y: usize,
    map: &mut HashMap<(usize, usize), i32>,
) -> i32 {
    if x == triangle.len() {
        return 0;
    }

    if map.contains_key(&(x, y)) {
        return *(map.get(&(x, y)).unwrap());
    }

    let left = divide_conquer(triangle, x + 1, y, map);
    let right = divide_conquer(triangle, x + 1, y + 1, map);
    let path = min(left, right) + triangle[x][y];
    map.insert((x, y), path);

    return path;
}
